provider/vsphere/internal/vsphereclient/createvm.go

```diff
@@ -402,24 +402,48 @@ func (c *Client) selectDatastore(
 	if err := c.client.Retrieve(ctx, refs, nil, &datastores); err != nil {
 		return nil, errors.Annotate(err, "retrieving datastore details")
 	}
+
+	if len(datastores) == 0 {
+		c.logger.Errorf("no datastores available")
```

Comment:
No need to log and return an error. The error will be logged.
Also, the message should include the requested datastore, eg
cannot use requested datastore "foo" because no datastores are available

Reply:
At this stage, we don't know whether the user has specified a desired datastore. I can add a test here and format accordingly, but it might make the code more complicated.